[Lens] Multi terms dnd support in Lens

dej611 opened the pull request:
## Summary

Build on top of #118600 

Drag and drop implementation for combine.

|  Scenario | GIF (click to zoom)  |
|---|---|
|  ✅  Drop field B into `Top Values of A` and combine into `Top Value of A > B` |   ![dnd_combine](https://user-images.githubusercontent.com/924948/143878904-d156b26a-bb5f-45d5-abce-08ee5e9fc2c6.gif) |
|  ✅  Drop field B into `Top Values of A` and combine into `Top Value of A > B` via keyboard  | ![dnd_combine_2](https://user-images.githubusercontent.com/924948/143878877-0158d9db-580d-41ac-bfcd-f0e52f13070b.gif)  |
|  ✅  Drop `Top Values of B` into `Top Values of A` and combine into `Top Values of A > B` | ![combine_dnd_drop_A_to_B](https://user-images.githubusercontent.com/924948/144582714-61e00105-c418-40d2-88b3-037ad880460d.gif)  |
|  ✅  Drop `Top Values of C` into `Top Values of A > B` and combine into `Top Values of A > B > C` |  ![combine_dnd_compatible_field_dimension](https://user-images.githubusercontent.com/924948/144582478-ad2d2beb-5189-40de-8053-4d91d83bbf8b.gif) |
|  ✅  Drop `Top Values of B` into `Top Values of A` and combine into `Top Values of A > B` via keyboard | ![combine_dnd_drop_A_to_B_keyboard](https://user-images.githubusercontent.com/924948/144582917-3d9db484-6a98-436a-873f-ad7569d82d78.gif)  |
|  ✅  Drop `Top Values of B > C` into `Top Values of A` and combine into `Top Values of A > B > C` | ![combine_dnd_drop_BC_to_A](https://user-images.githubusercontent.com/924948/144583177-dca3cf1f-6a3c-4251-979c-82a3428238ad.gif) |
|  ✅  Drop `Top Values of B > C` into `Top Values of A > B` and partially combine into `Top Values of A > B > C` | ![combine_dnd_drop_BC_to_AB](https://user-images.githubusercontent.com/924948/144583587-c06468b0-7372-40e5-bd6b-b8795f84b157.gif) |
|  ✅  Drop `Top Values of C > B` into `Top Values of A > B` and partially combine into `Top Values of A > B > C` | ![combine_dnd_drop_CB_to_AB](https://user-images.githubusercontent.com/924948/144583596-fa8e6ae2-6acf-40f8-81dd-67a7a0c3a5cd.gif) |
|  ✅  Drop `Unique count of B` into `Top Values of A` and combine into `Top Values of A > B` | ![combine_dnd_drop_metric_B_to_A](https://user-images.githubusercontent.com/924948/144589651-1e24a31a-6156-4a5d-91a4-4d0569ef5c1d.gif) |
|  ❌  Incompatible drop of `Top Values of A` into `Top Values of A` | ![combine_dnd_drop_same_dimension_A_to_A](https://user-images.githubusercontent.com/924948/144590142-9ff15c29-2865-41a7-a397-6150e30857fd.gif) |
|  ❌  No combine drop of `Top Values of A` into `Top Values of A > B` | ![combine_dnd_drop_same_dimension_A_to_AB](https://user-images.githubusercontent.com/924948/147767079-48bbe2f5-1000-45e4-99f7-179c15bc4347.gif) |
|  ❌  Incompatible drop of `Top Values of A > B` into `Top Values of A > B` | ![combine_dnd_drop_same_dimension_AB_to_AB](https://user-images.githubusercontent.com/924948/144590398-b3698094-92f8-4b94-93a2-314f4cf15698.gif) |
|  ❌  Incompatible drop of field `A` into `Top Values of A` | ![combine_dnd_drop_same_field_A_to_A](https://user-images.githubusercontent.com/924948/144590176-09ab8e8c-4951-409b-be9f-5999ef7d12b0.gif) |
|  ❌  Incompatible drop of `Unique count of A` into `Top Values of A` | ![combine_dnd_drop_metric_A_to_A](https://user-images.githubusercontent.com/924948/144589781-6b206aac-f253-4110-b9bf-054544c2e465.gif) |
|  ❌  Non-combinable drop of `Top Values of E` into `Top Values of A > B > C > D` for max limit of fields reached | ![combine_dnd_drop_same_dimension_E_to_ABCD](https://user-images.githubusercontent.com/924948/144590752-cedbe9b8-07f7-40a5-9ba9-62878b7eed85.gif) |
|  ❌  Non-combinable drop of `Top Values of D > E` into `Top Values of A > B > C` for max limit of fields reached | ![combine_dnd_drop_same_dimension_DE_to_ABC](https://user-images.githubusercontent.com/924948/144590863-5626c83d-02d6-458a-a93d-49eaaa2bfac7.gif) |

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@elasticmachine merge upstream

stratoula commented:
This looks great @dej611. The only thing that I found so far is that when I drag and drop from horizontal to break down, I see the 3 options. When I click control the `dupliccate`is highlighted and not the `Combine`. If I press Control it behaves ad duplicate and not as Combine

dej611 commented:
> This looks great @dej611. The only thing that I found so far is that when I drag and drop from horizontal to break down, I see the 3 options. When I click control the `dupliccate`is highlighted and not the `Combine`. If I press Control it behaves as duplicate and not as Combine

Good spot! Will fix ASAP.

dej611 commented:
@elasticmachine merge upstream

mbondyra commented on x-pack/plugins/lens/public/drag_drop/drag_drop.tsx:
```diff
@@ -274,11 +274,18 @@ const DragInner = memo(function DragInner({
   const modifierHandlers = useMemo(() => {
     const onKeyUp = (e: KeyboardEvent<HTMLButtonElement>) => {
-      if ((e.key === 'Shift' || e.key === 'Alt') && activeDropTarget?.id) {
+      if ((e.key === 'Shift' || e.key === 'Alt' || e.key === 'Control') && activeDropTarget?.id) {
```
nit: 
```suggestion
      if (activeDropTarget?.id && ['Shift', 'Alt', 'Control'].includes(e.key)) {
```

mbondyra reviewed (commented):
Once the functional tests pass, I can approve. Tested and couldn't find anything not working. There is one not-very important inconsistent behavior:
With the setup on X dimension Top Values of A + B and Split dimension Top Values of A, you can combine to split dimension, but you cannot combine to X dimension (you cannot do anything).

With the setup on X dimension Top Values of A + B and Split dimension Top Values of B, you can combine to split dimension, but you cannot combine to X dimension (but you can do other things).

https://user-images.githubusercontent.com/4283304/147673349-7aedbb65-ea41-4afb-999a-7acfaec3bb34.mp4


https://user-images.githubusercontent.com/4283304/147673370-5d1a99a5-dd8c-4652-8eef-291b9f0166f5.mp4

dej611 commented:
> With the setup on X dimension Top Values of A + B and Split dimension Top Values of A, you can combine to split dimension, but you cannot combine to X dimension (you cannot do anything).
> 
> With the setup on X dimension Top Values of A + B and Split dimension Top Values of B, you can combine to split dimension, but you cannot combine to X dimension (but you can do other things).

I haven't tested this scenario before. I understand from the user prospective fields in a multi terms have equivalent importance, so `duplicate` and `swap` should appear also for the latter.
I'll try to address this behaviour 👍

dej611 commented:
I think I've sorted out the remaining scenarios as well.

:x: Same operation with the same field (singular):
![same_field](https://user-images.githubusercontent.com/924948/147766793-3fb84244-ec63-4e91-a0c0-a93b332ef77a.gif)

:x: Same operation with same fields (multiple - order does not matter):

![same_fields](https://user-images.githubusercontent.com/924948/147766887-cd9d8e1d-2928-4383-a3ab-2fdc334d2c6c.gif)

✅  Same operation but different set of fields (A vs A+B) still allow swap, replace, and duplicate but not combine, the other direction has combine:

![same_subset_fields](https://user-images.githubusercontent.com/924948/147767079-48bbe2f5-1000-45e4-99f7-179c15bc4347.gif)

Added also more unit tests for these cases.

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@elasticmachine merge upstream

flash1293 commented:
@elasticmachine merge upstream

flash1293 reviewed (commented):
Looks mostly good to me, left a bunch of small things

dej611 commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/terms/index.tsx:
```diff
@@ -66,15 +66,64 @@ const idPrefix = htmlIdGenerator()();
+      ...(targetColumn.params?.secondaryFields ?? []),
+    ]);
+    // now check how many fields can be added
+    let counter = field && !originalTerms.has(field.name) ? 1 : 0;
```
The check is performed upstream, but we might include an additional check here just to be safe

dej611 commented:
@flash1293 sorted the scripted fields check and added more unit tests to that, on top of the other feedback fixes.
Had to tweak a bit some mocks to offer more fields (and took the opportunity to refactor).

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/terms/index.tsx:
```diff
@@ -62,16 +63,20 @@ function ofName(name?: string, count: number = 0) {
   });
 }
 
+function isScriptedField(field: IndexPatternField): boolean;
```
Is the overload necessary here? It seems like the last definition is capturing everything.

dej611 replied:
Initially I had only the generic third signature, but then thought it may be useful to express to the consumer why an `indexPattern` was required even when a field object was passed.
So the idea of the separate signatures which strongly express the relationship between the string name version with the `indexPattern`, compared to the other one.

Not particular strong opinion here, as the implementation is actually the same, just a type enhancement.

flash1293 replied:
ah got it, good point 👍

flash1293 reviewed (approved):
Tested again and still works, cool PR 👍
